### PR TITLE
avoid access to instructiont::guard

### PR DESCRIPTION
**Pull request by kroening**
This will eventually enable us to protect `instructiont::guard`.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/analyses/goto_check.cpp [hunk 1]
```diff
@@ -1668,12 +1677,18 @@ void goto_checkt::rw_ok_check(exprt &expr)
...
     exprt::operandst conjuncts;
+
     for(const auto &c : conditions)
       conjuncts.push_back(c.assertion);
 
-    expr = conjunction(conjuncts);
+    return conjunction(conjuncts);
   }
+  else if(modified)
+    return expr;
```
Comment (tautschnig) [inline src/analyses/goto_check.cpp]: `return std::move(expr);`

File: src/analyses/goto_check.cpp [hunk 2]
```diff
@@ -1831,7 +1846,9 @@ void goto_checkt::goto_check(
     {
       bool is_user_provided=i.source_location.get_bool("user-provided");
 
-      rw_ok_check(i.guard);
+      auto rw_ok_cond = rw_ok_check(i.get_condition());
+      if(rw_ok_cond.has_value())
+        i.set_condition(*rw_ok_cond);
```
Comment (tautschnig) [inline src/analyses/goto_check.cpp:1851]: Shouldn't this be a case for `i.transform([this](exprt e){ return rw_ok_check(e); });`?

File: src/goto-instrument/accelerate/scratch_program.cpp [hunk 3]
```diff
@@ -162,12 +164,12 @@ void scratch_programt::append_path(patht &path)
...
       if(it->guard.id()!=ID_nil)
       {
-        add_instruction(ASSUME)->guard=it->guard;
+        add_instruction(ASSUME)->set_condition(it->guard);
       }
     }
     else if(it->loc->is_assert())
     {
-      add_instruction(ASSUME)->guard=it->loc->guard;
+      add_instruction(ASSUME)->set_condition(it->loc->get_condition());
```
Comment (tautschnig) [inline src/goto-instrument/accelerate/scratch_program.cpp]: `add(goto_programt::make_assume(it->loc->get_condition()));`

File: src/goto-instrument/accelerate/scratch_program.cpp [hunk 4]
```diff
@@ -162,12 +164,12 @@ void scratch_programt::append_path(patht &path)
     {
       if(it->guard.id()!=ID_nil)
       {
-        add_instruction(ASSUME)->guard=it->guard;
+        add_instruction(ASSUME)->set_condition(it->guard);
```
Comment (tautschnig) [inline src/goto-instrument/accelerate/scratch_program.cpp]: `add(goto_programt::make_assume(it->get_condition()));`

File: src/goto-instrument/dot.cpp [hunk 5]
```diff
@@ -180,7 +180,9 @@ void dott::write_dot_subgraph(
 
       out << "Node_" << subgraphscount << "_" << it->location_number;
       out << " [shape=";
-      if(it->is_goto() && !it->guard.is_true() && !it->guard.is_false())
+      if(
+        it->is_goto() && !it->get_condition().is_true() &&
+        !it->get_condition().is_false())
```
Comment (tautschnig) [inline src/goto-instrument/dot.cpp]: How about `!it->get_condition().is_constant()`?

File: src/goto-instrument/nondet_volatile.cpp [hunk 6]
```diff
@@ -83,7 +83,7 @@ void nondet_volatile(
 
   Forall_goto_program_instructions(i_it, goto_program)
   {
-    goto_programt::instructiont &instruction=*i_it;
+    goto_programt::instructiont &instruction = *i_it;
```
Comment (tautschnig) [inline src/goto-instrument/nondet_volatile.cpp]: Unnecessary whitespace change?

File: src/goto-instrument/wmm/shared_buffers.cpp [hunk 7]
```diff
@@ -320,8 +320,8 @@ void shared_bufferst::write(
     not_exprt(and_exprt(buff1_used_expr, buff0_used_expr));
 
   target=goto_program.insert_before(target);
-  target->guard=cond_expr;
   target->type=ASSERT;
+  target->set_condition(cond_expr);
   target->code = code_assertt(cond_expr);
   target->code.add_source_location()=source_location;
```
Comment (tautschnig) [inline src/goto-instrument/wmm/shared_buffers.cpp]: Should be rewritten using `goto_programt::make_assert` (as is already done elsewhere in this commit).

File: src/goto-programs/set_properties.cpp [hunk 8]
```diff
@@ -150,7 +150,7 @@ void make_assertions_false(
     {
       if(!i_it->is_assert())
         continue;
-      i_it->guard=false_exprt();
+      i_it->set_condition(false_exprt());
```
Comment (tautschnig) [inline src/goto-programs/set_properties.cpp]: Asking for an unrelated change: can we please write this as
```
if(i_it->is_assert())
  i_it->set_condition(false_exprt());
```

**Review by tautschnig — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 2924a3c744123668d131de93566a4ef5e214cd91).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/104193257

File: src/analyses/goto_check.cpp (hunk 1 above)
Comment (kroening) [reply src/analyses/goto_check.cpp]: done

File: src/analyses/goto_check.cpp (hunk 2 above)
Comment (kroening) [reply src/analyses/goto_check.cpp:1851]: Yes, will do separate PR.

File: src/goto-instrument/accelerate/scratch_program.cpp (hunk 4 above)
Comment (kroening) [reply src/goto-instrument/accelerate/scratch_program.cpp]: Ok, made that a separate commit.

File: src/goto-instrument/dot.cpp (hunk 5 above)
Comment (kroening) [reply src/goto-instrument/dot.cpp]: sure

File: src/goto-instrument/nondet_volatile.cpp (hunk 6 above)
Comment (kroening) [reply src/goto-instrument/nondet_volatile.cpp]: removed

File: src/goto-programs/set_properties.cpp (hunk 8 above)
Comment (kroening) [reply src/goto-programs/set_properties.cpp]: sure

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: b604f1bab62d3f8945f902e69a6689e06a25a5c5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/104199319
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e44f5ee3b2755484392cd14efb281042f78e95d7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/104202022

**Review by smowton — APPROVED** [review]

File: src/goto-analyzer/static_simplifier.cpp [hunk 9]
```diff
@@ -65,33 +65,45 @@ bool static_simplifier(
 
       if(i_it->is_assert())
       {
-        bool unchanged =
-          ai.abstract_state_before(i_it)->ai_simplify(i_it->guard, ns);
+        exprt cond = i_it->get_condition();
+
+        bool unchanged = ai.abstract_state_before(i_it)->ai_simplify(cond, ns);
```
Comment (peterschrammel) [inline src/goto-analyzer/static_simplifier.cpp:70]: `const` (several occurrences)

**Review by peterschrammel — APPROVED** [review]
